<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="base_frame" default="base_link"/>

  <arg name="rviz" default="true"/>

  <arg name="lost_frame_max_acceleration" default="10.0"/>
  <arg name="calibration_skip_keyframes" default="7"/>
  <arg name="mapping_registrator" default="ndt" description="ndt or gicp"/>

  <let name="calibrate_base_frame" value="true"/>
  <arg name="base_lidar_crop_box_min_x" default="-5.0"/>
  <arg name="base_lidar_crop_box_min_y" default="-5.0"/>
  <arg name="base_lidar_crop_box_min_z" default="-5.0"/>
  <arg name="base_lidar_crop_box_max_x" default="10.0"/>
  <arg name="base_lidar_crop_box_max_y" default="5.0"/>
  <arg name="base_lidar_crop_box_max_z" default="5.0"/>
  <arg name="base_lidar_min_plane_points_percentage" default="10.0"/>
  <arg name="base_lidar_max_inlier_distance" default="0.03"/>
  <arg name="base_lidar_min_plane_points" default="500"/>
  <arg name="base_lidar_max_cos_distance" default="0.2"/>
  <arg name="base_lidar_max_iterations" default="500"/>
  <arg name="base_lidar_overwrite_xy_yaw" default="false"/>

  <let name="calibration_camera_optical_link_frames" value="['']"/>

  <let name="calibration_lidar_frames" value="['']"/>

  <let name="mapping_lidar_frame" value="hesai_top"/>
  <let name="mapping_pointcloud" value="/sensing/lidar/top/pointcloud"/>
  <let name="detected_objects" value="/perception/object_recognition/detection/objects"/>

  <let name="calibration_camera_info_topics" value="['']"/>

  <let name="calibration_image_topics" value="[
    '']"/>

  <let name="calibration_pointcloud_topics" value="['']"/>

  <!-- mapping based calibrator -->
  <include file="$(find-pkg-share mapping_based_calibrator)/launch/calibrator.launch.xml">
    <arg name="ns" value=""/>
    <arg name="calibration_service_name" value="calibrate_base_lidar"/>

    <arg name="rviz" value="$(var rviz)"/>
    <arg name="base_frame" value="$(var base_frame)"/>

    <arg name="calibration_camera_optical_link_frames" value="$(var calibration_camera_optical_link_frames)"/>
    <arg name="calibration_lidar_frames" value="$(var calibration_lidar_frames)"/>
    <arg name="mapping_lidar_frame" value="$(var mapping_lidar_frame)"/>

    <arg name="mapping_pointcloud" value="$(var mapping_pointcloud)"/>
    <arg name="detected_objects" value="$(var detected_objects)"/>

    <arg name="calibration_camera_info_topics" value="$(var calibration_camera_info_topics)"/>
    <arg name="calibration_image_topics" value="$(var calibration_image_topics)"/>
    <arg name="calibration_pointcloud_topics" value="$(var calibration_pointcloud_topics)"/>

    <arg name="mapping_registrator" value="$(var mapping_registrator)"/>
    <arg name="lost_frame_max_acceleration" value="$(var lost_frame_max_acceleration)"/>
    <arg name="calibration_skip_keyframes" value="$(var calibration_skip_keyframes)"/>

    <arg name="base_lidar_crop_box_min_x" value="$(var base_lidar_crop_box_min_x)"/>
    <arg name="base_lidar_crop_box_min_y" value="$(var base_lidar_crop_box_min_y)"/>
    <arg name="base_lidar_crop_box_min_z" value="$(var base_lidar_crop_box_min_z)"/>
    <arg name="base_lidar_crop_box_max_x" value="$(var base_lidar_crop_box_max_x)"/>
    <arg name="base_lidar_crop_box_max_y" value="$(var base_lidar_crop_box_max_y)"/>
    <arg name="base_lidar_crop_box_max_z" value="$(var base_lidar_crop_box_max_z)"/>
    <arg name="base_lidar_min_plane_points_percentage" value="$(var base_lidar_min_plane_points_percentage)"/>
    <arg name="base_lidar_max_inlier_distance" value="$(var base_lidar_max_inlier_distance)"/>
    <arg name="base_lidar_min_plane_points" value="$(var base_lidar_min_plane_points)"/>
    <arg name="base_lidar_max_cos_distance" value="$(var base_lidar_max_cos_distance)"/>
    <arg name="base_lidar_max_iterations" value="$(var base_lidar_max_iterations)"/>
    <arg name="base_lidar_overwrite_xy_yaw" value="$(var base_lidar_overwrite_xy_yaw)"/>
  </include>
</launch>
